mapiproxy/servers/default/asyncemsmdb/dcesrv_asyncemsmdb.c

```diff
@@ -0,0 +1,1230 @@
+			*r->out.pulFlagsOut = 0x1;
+			return NT_STATUS_OK;
+		}
+		session->uuid = r->in.async_handle->uuid;
+		session->data = talloc_steal(session, p);
+		session->cn = talloc_strdup(session, cn);
+		talloc_free(cn);
+		if (!session->cn) {
+			talloc_free(session);
+			goto failure;
+		}
+		session->bind_addr = talloc_strdup(session, bind_addr);
+		talloc_free(bind_addr);
+		if (!session->bind_addr) {
+			talloc_free(session);
+			goto failure;
+		}
+		session->mstore_ctx = mstore_ctx;
+
+		/* Add the session to the dcesrv_connection_context */
+		dce_call->context->private_data = session;
+
+		OC_DEBUG(5, "[asyncemsmdb]: New session added: %s", session->data->emsmdb_session_str);
+		DLIST_ADD_END(asyncemsmdb_session, session, struct exchange_asyncemsmdb_session *);
+	}
+
+	return NT_STATUS_OK;
+}
+
+static NTSTATUS dcerpc_server_asyncemsmdb_unbind(struct dcesrv_connection_context *context, const struct dcesrv_interface *iface)
+{
+	enum mapistore_error			retval;
+	struct exchange_asyncemsmdb_session	*session = (struct exchange_asyncemsmdb_session *) context->private_data;
+
+	OC_DEBUG(0, "dcerpc_server_asyncemsmdb_unbind");
+
+	if (!session) {
+		return NT_STATUS_OK;
+	}
+
+	OC_DEBUG(0, "[asyncemsmdb] unbind %s", session->bind_addr);
+	retval = mapistore_notification_resolver_delete(session->mstore_ctx, session->cn, session->bind_addr);
+	if (retval != MAPISTORE_SUCCESS) {
+		OC_DEBUG(0, "[asyncemsmdb] unable to delete resolver entry %s from record %s", session->bind_addr, session->cn);
+	}
+
+	mapistore_release(session->mstore_ctx);
+	DLIST_REMOVE(asyncemsmdb_session, session);
+
+	/* flush pending call on connection */
+	context->conn->pending_call_list = NULL;
+
+	return NT_STATUS_OK;
+}
+
+
+/**
+   \details Initialize ldb_context to samdb, creates one for all emsmdbp
+   contexts
+
+   \param lp_ctx pointer to the loadparm context
+ */
+static struct ldb_context *samdb_init(struct loadparm_context *lp_ctx)
+{
+	TALLOC_CTX		*mem_ctx;
+	struct tevent_context	*ev;
+	const char		*samdb_url;
+	struct ldb_context	*_samdb_ctx;
+
+	if (samdb_ctx) return samdb_ctx;
+
+	mem_ctx = talloc_autofree_context();
+	ev = tevent_context_init(mem_ctx);
+	if (!ev) {
+		return NULL;
+	}
+	tevent_loop_allow_nesting(ev);
+
+	/* Retrieve samdb url (local or external) */
+	samdb_url = lpcfg_parm_string(lp_ctx, NULL, "dcerpc_mapiproxy", "samdb_url");
+
+	if (!samdb_url) {
+		_samdb_ctx = samdb_connect(mem_ctx, ev, lp_ctx, system_session(lp_ctx), 0);
+	} else {
+		_samdb_ctx = samdb_connect_url(mem_ctx, ev, lp_ctx, system_session(lp_ctx),
+					       LDB_FLG_RECONNECT, samdb_url);
+	}
+
+	return _samdb_ctx;
+}
+
+
+static NTSTATUS dcerpc_server_asyncemsmdb_bind(struct dcesrv_call_state *dce_call, const struct dcesrv_interface *iface)
+{
+	if (!openchangedb_ctx) {
+		openchangedb_ctx = mapiproxy_server_openchangedb_init(dce_call->conn->dce_ctx->lp_ctx);
+		if (!openchangedb_ctx) {
+			OC_DEBUG(OC_LOG_FATAL, "Unable to initialize openchangedb");
+			abort();
```

Comment:
dcesrv_asyncemsmdb links against libmapiproxy, it should therefore be straight forward to replace. change in progress